OADP-227: Which K8s API version is used; how to use Enable API Group Versions

RichardHoch opened the pull request:
OADP 1.1.0; OCP 4.9+

Resolves: https://issues.redhat.com/browse/OADP-227

Preview: http://file.emea.redhat.com/rhoch/api_group_versions/backup_and_restore/application_backup_and_restore/oadp-advanced-topics.html

Please do not merge before Sept. 1.

mperetzred commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,23 @@
+
+For example, if you have a resource that has 3 `apiVersions`  - `v1alpha1`, `v1beta1`, and `v1` - registered in the cluster, Velero only backs up the version of that resource that uses the `preferredVersion` of that CRD, v1. So, when Velero attempts to restore this resource, the cluster must have `v1` registered in its set of available `api-resources`.
+
+Therefore, you must check which Kubernetes API group versions you have on your source cluster.
```
I think it should be:
Therefore, you must check which Kubernetes API group versions you have on your **target** cluster.

RichardHoch replied:
Done.

mperetzred commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,23 @@
+
+If you use Velero to backup and restore a source cluster that has a resource that has multiple Kubernetes API versions registered to it, Velero backs up only the version of that resource that uses the `preferredVersion` of that resource's Kubernetes API.
+
+For example, if you have a resource that has 3 `apiVersions`  - `v1alpha1`, `v1beta1`, and `v1` - registered in the cluster, Velero only backs up the version of that resource that uses the `preferredVersion` of that CRD, v1. So, when Velero attempts to restore this resource, the cluster must have `v1` registered in its set of available `api-resources`.
```
I think it should be:
So, when Velero attempts to restore this resource, the **target** cluster must have `v1` registered in its set of available `api-resources`.

RichardHoch replied:
@mperetzred When you run `oc api-resources` does it return information for your source cluster, your target cluster, or both?

mperetzred replied:
it depends which cluster you are currently connected to on your terminal, like any oc command.

RichardHoch replied:
Done.

RichardHoch commented:
@mperetzred I made the changes you suggested. Please review.
@dymurray Please review this PR.

RichardHoch commented:
@sbeskin-redhat: Please review.

sbeskin-redhat commented on backup_and_restore/application_backup_and_restore/oadp-advanced-topics.adoc:
```diff
@@ -0,0 +1,21 @@
+:oadp-advanced-topics:
+:namespace: openshift-adp
+:local-product: OADP
+:velero-domain: velero.io
```
@RichardHoch 
This will not be necessary after OADP-563 is merged.

sbeskin-redhat commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,23 @@
+[id="oadp-checking-api-group-versions_{context}"]
+= Checking Kubernetes API group versions
+
+If you use Velero to backup and restore a source cluster that has a resource that has multiple Kubernetes API versions registered to it, Velero backs up only the version of that resource that uses the `preferredVersion` of that resource's Kubernetes API.
```
@RichardHoch 
"...to back up..."
Also, this is a very cumbersome sentence. I would suggest splitting it into a few shorter ones. Something like this:
"A source cluster might have a resource with multiple Kubernetes API versions registered to it. One of those versions is the preferred version. If you use Velero to back up and restore such a source cluster, Velero backs up only the version of that resource that uses the preferred version of its Kubernetes API."

sbeskin-redhat commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,23 @@
+
+If you use Velero to backup and restore a source cluster that has a resource that has multiple Kubernetes API versions registered to it, Velero backs up only the version of that resource that uses the `preferredVersion` of that resource's Kubernetes API.
+
+For example, if you have a resource that has 3 `apiVersions`  - `v1alpha1`, `v1beta1`, and `v1` - registered in the cluster, Velero only backs up the version of that resource that uses the `preferredVersion` of that CRD, v1. So, when Velero attempts to restore this resource, the target cluster must have `v1` registered in its set of available `api-resources`.
```
@RichardHoch 
We are not supposed to treat backticked object names as regular nouns. So:
"... 3 API versions..."
"...the preferred API version of that CRD"
"...must have the `v1` version..."
"... available API resources."
Also:
"In the cluster" -> "On the cluster"? I think we use 'on' with cluster, like you did below.

sbeskin-redhat commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+[id="oadp-using-enable-api-group-versions_{context}"]
+= Working with different Kubernetes API versions in the same OADP operation
+
+By default, Velero only backs up resources that use the `preferredAPI` Kubernetes version. However, Velero also includes a feature, link:https://velero.io/docs/v1.9/enable-api-group-versions-feature/[Enable API Group Versions], that overcomes this limitation. When enabled on the source cluster, this feature causes Velero to back up _all_ supported Kubernetes API group versions on the cluster, not ...
```
@RichardHoch 
"...use the preferred Kubernetes API version."
I would also suggest:
"..._all_  Kubernetes API group versions that are supported on the cluster..."

sbeskin-redhat commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+
+[NOTE]
+====
+As of August 2022, Enable API is still in beta.
```
@RichardHoch 
"...2022, Enable API Group Versions is still a beta feature."

sbeskin-redhat commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+As of August 2022, Enable API is still in beta.
+====
+
+For example, consider the resource `horizontalpodautoscalers`, which falls under the `autoscaling` group and has 3 `apiVersions` - `v1`,`v2beta1`, and `v2beta2` - registered in the cluster, with the preferred API being `v1`.
```
@RichardHoch 
"...has 3 API versions..."
"...on the cluster..."

sbeskin-redhat commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+
+For example, consider the resource `horizontalpodautoscalers`, which falls under the `autoscaling` group and has 3 `apiVersions` - `v1`,`v2beta1`, and `v2beta2` - registered in the cluster, with the preferred API being `v1`.
+
+Without the Enable API Group Versions enabled, only the preferred API group version for `autoscaling`, `v1`, is backed up. With the feature enabled, the remaining supported versions, `v2beta1` and `v2beta2` are also backed up. Once the versions are stored in the backup tarball file, they will be available to be restored on the destination cluster.
```
@RichardHoch 
"Without the Enable API Group Versions feature enabled, Velero backs up only the preferred API group version for the `autoscaling` group, `v1`. With the feature enabled, Velero backs up the remaining supported versions, `v2beta1` and `v2beta2`."
"...in the backup tarball file, they are available to be restored..."

sbeskin-redhat commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+
+Without the Enable API Group Versions enabled, only the preferred API group version for `autoscaling`, `v1`, is backed up. With the feature enabled, the remaining supported versions, `v2beta1` and `v2beta2` are also backed up. Once the versions are stored in the backup tarball file, they will be available to be restored on the destination cluster.
+
+When the Enable API Group Versions feature is enabled on the destination cluster, Velero restore will choose the version to restore based on an API group version priority order.
```
@RichardHoch 
"...on the destination cluster, Velero selects the version to restore on the basis of the order of priority of API group versions."

sbeskin-redhat commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+
+Velero uses the following algorithm to assign priorities to API versions, with 1 as the top priority:
+
+Priority 1: Destination cluster preferred version +
```
@RichardHoch 
"...cluster's..."

sbeskin-redhat commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+Velero uses the following algorithm to assign priorities to API versions, with 1 as the top priority:
+
+Priority 1: Destination cluster preferred version +
+Priority 2: Source cluster preferred version +
```
@RichardHoch 
"...cluster's..."

jboxman-rh commented on _topic_maps/_topic_map.yml:
```diff
@@ -2379,6 +2379,8 @@ Topics:
       File: restoring-applications
   - Name: Troubleshooting
     File: troubleshooting
+  - Name: Advanced Topics
```
Can we be more description? Also, we use sentence case for titles here.

jboxman-rh commented on backup_and_restore/application_backup_and_restore/oadp-advanced-topics.adoc:
```diff
@@ -0,0 +1,20 @@
+= Advanced topics
+include::_attributes/common-attributes.adoc[]
+:context: oadp-advanced-topics
+:oadp-advanced-topics:
```
Do we need this empty attribute?

jboxman-rh commented on backup_and_restore/application_backup_and_restore/oadp-advanced-topics.adoc:
```diff
@@ -0,0 +1,20 @@
+include::_attributes/common-attributes.adoc[]
+:context: oadp-advanced-topics
+:oadp-advanced-topics:
+:namespace: openshift-adp
```
This doesn't seem to be used anywhere?

jboxman-rh commented on backup_and_restore/application_backup_and_restore/oadp-advanced-topics.adoc:
```diff
@@ -0,0 +1,20 @@
+:context: oadp-advanced-topics
+:oadp-advanced-topics:
+:namespace: openshift-adp
+:local-product: OADP
```
This doesn't seem to be used anywhere?

jboxman-rh commented on backup_and_restore/application_backup_and_restore/oadp-advanced-topics.adoc:
```diff
@@ -0,0 +1,20 @@
+
+toc::[]
+
+This document provides information on advanced features and functionalities of OADP.
```
What's an OADP?

jboxman-rh replied:
Looks like it is: OpenShift API for Data Protection Operator

jboxman-rh commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,23 @@
+
+:_content-type: PROCEDURE
+[id="oadp-checking-api-group-versions_{context}"]
+= Checking Kubernetes API group versions
```
How about

> Listing the Kubernetes APIs available in the cluster

RichardHoch replied:
@jboxman-rh -- How about "Listing the Kubernetes API group versions in a cluster" ?

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+
+[NOTE]
+====
+As of August 2022, Enable API Group Versions is still in beta.
```
We generally avoid dates and times, as this boxes us in as a company. How about:

> Enable API Group Versions is still in beta.

jboxman-rh commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,23 @@
+[id="oadp-checking-api-group-versions_{context}"]
+= Checking Kubernetes API group versions
+
+A source cluster might have a resource with multiple Kubernetes API versions registered to it. One of those versions is the preferred version. If you use Velero to back up and restore such a source cluster, Velero backs up only the version of that resource that uses the preferred version of its Kubernetes API.
```
So in k8s there might be multiple versions of an API available. So the causality is backwards here, I think. Instead:

> A source cluster might offer multiple versions of an API, where one of these versions is the preferred API version. For example, a source cluster with an API named `Example` might be available in the `example.com/v1` and `example.com/v1beta2` API groups.

Something like that, at any rate.

The key is that only API objects created against these preferred APIs are backed up. For more particulars, there is this: https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/

But we're talking about these:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
spec:
...
```

jboxman-rh commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,23 @@
+
+A source cluster might have a resource with multiple Kubernetes API versions registered to it. One of those versions is the preferred version. If you use Velero to back up and restore such a source cluster, Velero backs up only the version of that resource that uses the preferred version of its Kubernetes API.
+
+For example, if you have a resource that has 3 API versions - `v1alpha1`, `v1beta1`, and `v1` - registered on the cluster, Velero only backs up the version of that resource that uses the preferred API version of that CRD, v1. So, when Velero attempts to restore this resource, the target cluster must have the `v1` version registered in its set of available API resources.
```
It might not be a CRD. A custom resource definition is actually an extension of the k8s API. Meanwhile, k8s has certain built-in APIs, like Pod, that are not defined by a CRD. So while APIs in OCP are generally (always?) based on a CRD, many k8s built-ins are _not_.

jboxman-rh replied:
It's not really a resource that _has_ an API, exactly. Rather, the cluster offers different versions of an API, and that API lets you create some kind of resource.

jboxman-rh commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,23 @@
+
+For example, if you have a resource that has 3 API versions - `v1alpha1`, `v1beta1`, and `v1` - registered on the cluster, Velero only backs up the version of that resource that uses the preferred API version of that CRD, v1. So, when Velero attempts to restore this resource, the target cluster must have the `v1` version registered in its set of available API resources.
+
+Therefore, you must check which Kubernetes API group versions you have on your target cluster.
```
```suggestion
Therefore, you must verify which Kubernetes API group versions you have on your target cluster.
```
Or _confirm_ perhaps.

RichardHoch replied:
I believe the intent is that you don't necessarily know which versions you have. How about "You need to generate a list of the Kubernetes..." ?

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+As of August 2022, Enable API Group Versions is still in beta.
+====
+
+For example, consider the resource `horizontalpodautoscalers`, which falls under the `autoscaling` group and has 3 API versions - `v1`,`v2beta1`, and `v2beta2` - registered on the cluster, with the preferred API being `v1`.
```
So technically this is the _plural_ for that. The actual API is called `HorizontalPodAutoscaler`. See the API docs: https://docs.openshift.com/container-platform/4.10/rest_api/autoscale_apis/horizontalpodautoscaler-autoscaling-v1.html#horizontalpodautoscaler-autoscaling-v1

The API group here is `autoscaling/v1`.

The plural is actually related to the REST API URL. I've always found that confusing, but it is this way _because REST_.

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+
+Velero uses the following algorithm to assign priorities to API versions, with 1 as the top priority:
+
+Priority 1: Preferred version of the _destination_ cluster+
```
There are stray + here.

jboxman-rh replied:
Would this make more sense as an ordered list?

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+
+.Procedure
+
+* Configure the `EnableAPIGroupVersions` feature flag:
```
Configure this _where_? What is this `spec` related to? In what API is this from? Every API in OCP almost without exception has a `spec` field, so this by itself doesn't help me understand where this takes place.

RichardHoch replied:
@kaovilai @dymurray -- can you help here?

kaovilai replied:
Configure right here.. this spec is part of DataProtectionApplication spec
https://github.dev/RichardHoch/openshift-docs/blob/9f03e415c9fe53cdbfcfe095eb482cf1f0dbb05c/modules/oadp-using-enable-api-group-versions.adoc#L35-L39

Dpa spec is first described here
https://github.dev/RichardHoch/openshift-docs/blob/9f03e415c9fe53cdbfcfe095eb482cf1f0dbb05c/modules/oadp-installing-dpa.adoc#L41-L80

kaovilai replied:
This API is part of the feature flags parameter for velero which is an option you can configure in DPA.

RichardHoch replied:
@kaovilai 

So woud the full yaml would be as follows?

apiVersion: oadp.openshift.io/v1alpha1
kind: DataProtectionApplication
----
----
----
spec:
  configuration:
    velero:
      featureFlags:
      - EnableAPIGroupVersions

jboxman-rh replied:
@kaovilai, do we know what namespace this belongs in when it's created, and what the proper `name` is for the API object? Is it user-supplied or is there an canonical name like for `config.openshift.io` APIs? Thanks!

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,44 @@
+
+[role="_additional-resources"]
+.Additional resources
+link:https://velero.io/docs/v1.9/enable-api-group-versions-feature/[Enable API Group Versions Feature]
```
I think we need a single bullet here for this

jboxman-rh reviewed (commented):
@RichardHoch, I've left some comments for your review. Let me know if you have any questions, thanks!

RichardHoch commented:
@jboxman-rh I made the changes you suggested -- for some reason, i cannot export my local build using 
rsync -avz --progress _preview/openshift-enterprise/api_group_versions/* rhoch@file.emea.redhat.com:/home/tlv/rhoch/public_html/api_group_versions/

as usual, so the preview is messed up.

Could you please look at the files nonetheless and perhaps advise why the command that always works suddenly isn't? 
I need to leave now, so take your time.

jboxman-rh commented:
> @jboxman-rh I made the changes you suggested -- for some reason, i cannot export my local build using rsync -avz --progress _preview/openshift-enterprise/api_group_versions/* [rhoch@file.emea.redhat.com](mailto:rhoch@file.emea.redhat.com):/home/tlv/rhoch/public_html/api_group_versions/
> 
> as usual, so the preview is messed up.
> 
> Could you please look at the files nonetheless and perhaps advise why the command that always works suddenly isn't? I need to leave now, so take your time.

Out of disk space maybe? It doesn't provide any useful output or errors?

RichardHoch commented:
No info or sign of a problem -- I am doing the same thing i have been doing all day with all my other PRs. I noticed an odd thing in the topics map file -- looks like i removed a line and added it but i remember doing so such thing. Maybe take a look?

RichardHoch commented:
@jboxman-rh Oh, this is "interesting"

rhoch@rhoch openshift-docs (api_group_versions) $ rsync -avz --progress _preview/openshift-enterprise/api_group_versions/* rhoch@file.emea.redhat.com:/home/tlv/rhoch/public_html/api_group_versions/
ssh: Could not resolve hostname file.emea.redhat.com: Name or service not known
rsync: connection unexpectedly closed (0 bytes received so far) [sender]
rsync error: unexplained error (code 255) at io.c(226) [sender=3.1.3]

Never saw that before... 
Thoughts?

kaovilai commented:
> ssh: Could not resolve hostname file.emea.redhat.com: Name or service not known
Sounds like you got disconnected from vpn.. or unstable internet.

kaovilai commented:
Problem with `Could not resolve hostname` typically are DNS issues. In this case it did not resolve this address at redhat.com which I assume is an internal link.

RichardHoch commented:
@jboxman-rh I made almost all of the changes you recommended, suggesting an alternative to one. I also got the preview to work again. Please review this PR again. Thanks.

RichardHoch commented:
@sbeskin-redhat: I made the changes you suggested. Please review this PR again.

sbeskin-redhat commented:
@RichardHoch 
Looks fine

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+Enable API Group Versions is still in beta.
+====
+
+Velero uses the following algorithm to assign priorities to API versions, with 1 as the top priority:
```
```suggestion
Velero uses the following algorithm to assign priorities to API versions, with `1` as the top priority:
```

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+
+Velero uses the following algorithm to assign priorities to API versions, with 1 as the top priority:
+
+. Preferred version of the _destination_ cluster+
```
```suggestion
. Preferred version of the _destination_ cluster
```

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+Velero uses the following algorithm to assign priorities to API versions, with 1 as the top priority:
+
+. Preferred version of the _destination_ cluster+
+. Preferred version of the source_ cluster  +
```
```suggestion
. Preferred version of the source_ cluster
```

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+
+* Configure the `EnableAPIGroupVersions` feature flag:
+
+[source, yaml]
```
```suggestion
[source,yaml]
```

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
````diff
@@ -0,0 +1,49 @@
+```
+apiVersion: oadp.openshift.io/vialpha1
+kind: DataProtectionApplication
+...
````
See this comment https://github.com/openshift/openshift-docs/pull/48202/files#r957990388 for my questions about this API.

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+. Preferred version of the source_ cluster  +
+. Common non-preferred supported version with the highest Kubernetes version priority
+
+.Procedure
```
So this one step procedure doesn't by itself actually enable the feature flag; The user needs to perform some action with this, such as `oc apply -f <my.yaml>`.

There's more to this story, and we need to tell it.

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+
+:_content-type: PROCEDURE
+[id="oadp-using-enable-api-group-versions_{context}"]
+= Working with different Kubernetes API versions in the same OADP operation
```
What does it mean to have these happen in the same _operation_. Is this the backup? Could it be backup or restore? And _operation_ is thus a superset of the possibilities here?

kalexand-rh replied:
I agree - "operation" is not a clear word choice here.

jboxman-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+[id="oadp-using-enable-api-group-versions_{context}"]
+= Working with different Kubernetes API versions in the same OADP operation
+
+By default, Velero only backs up resources that use the preferred version of the Kubernetes API. However, Velero also includes a feature, link:https://velero.io/docs/v1.9/enable-api-group-versions-feature/[Enable API Group Versions], that overcomes this limitation. When enabled on the source cluster, this feature causes Velero to back up _all_ Kubernetes API group versions that are supported on t ...
```
@kaovilai, I thought API objects tended to be upgraded automatically to the latest version; A cluster can have API objects that are against an older API persistently?

I'm just trying to better understand when this situation comes into play.

Thanks!

kaovilai replied:
It's not automatic. If your yaml uses a specific version, it will use that version and fail if it doesn't exist in the cluster. Cluster can maintain multiple valid versions at any given point in time.

kaovilai replied:
For kube native or some controllers it's possible they do upgrade behind the scenes.. but if you oc create v1alpha1 I wouldn't expect everything with v1 to get upgraded to v1.

There could've been many changes between the two and automatic upgrade on everything would be problematic.

jboxman-rh reviewed (commented):
@RichardHoch, I left some additional comments, thanks!

jboxman-rh commented:
While there's room for some additional clarity in this, it's gotten SME sign-off and been through a peer review, so given the approaching deadline for this I'm going to say LGTM and we'll iterate on this some more in a future PR.

RichardHoch commented:
@jboxman-rh Thanks and ack.

kalexand-rh commented on backup_and_restore/application_backup_and_restore/oadp-advanced-topics.adoc:
```diff
@@ -0,0 +1,18 @@
+
+This document provides information on advanced features and functionalities of OpenShift API for Data Protection (OADP).
+
+[id="oadp-different-kubernetes-api-versions_{context}"]
```
[Anchors](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-in-assembly-files) in assemblies do not use the `_{context}` element.

kalexand-rh commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,25 @@
+
+If you use Velero to back up and restore such a source cluster, Velero backs up only the version of that resource that uses the preferred version of its Kubernetes API.
+
+In our case, if `example.com/v1` were the preferred API, then Velero would only back up the version of a resource that uses `example.com/v1`. Moreover, the target cluster would need to have `example.com/v1` registered in its set of available API resources in order for Velero to restore the resource on the target cluster.
```
Do not use "our." This isn't one of the situations where the ISG allows personal pronouns.

Also, please don't use "would" here.
> ! Would v
Use only to refer to an unfulfilled condition or hypothesis and to express the future in the past, for example, "The command was rejected because it would have resulted in the ClusterName attribute being blank." Do not use as an auxiliary verb, for example, "You would need administrator authority to create an app." Instead, use the present tense, for example "You need administrator authority to create an app."

RichardHoch replied:
Fixed; present simple used instead.

kalexand-rh commented on modules/oadp-checking-api-group-versions.adoc:
```diff
@@ -0,0 +1,25 @@
+
+[source,terminal]
+----
+oc api-resources
```
All [commands](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) must be prefixed with `$` (or `#` if root permissions are required).

kalexand-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+[id="oadp-using-enable-api-group-versions_{context}"]
+= Working with different Kubernetes API versions in the same OADP operation
+
+By default, Velero only backs up resources that use the preferred version of the Kubernetes API. However, Velero also includes a feature, link:https://velero.io/docs/v1.9/enable-api-group-versions-feature/[Enable API Group Versions], that overcomes this limitation. When enabled on the source cluster, this feature causes Velero to back up _all_ Kubernetes API group versions that are supported on t ...
```
> ! tarball noun
Do not use. Use ".tar file".

> ! once adverb
Use only to mean "one time". Do not use as a conjunction to mean "after" or "when".

RichardHoch replied:
Done

kalexand-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+[id="oadp-using-enable-api-group-versions_{context}"]
+= Working with different Kubernetes API versions in the same OADP operation
+
+By default, Velero only backs up resources that use the preferred version of the Kubernetes API. However, Velero also includes a feature, link:https://velero.io/docs/v1.9/enable-api-group-versions-feature/[Enable API Group Versions], that overcomes this limitation. When enabled on the source cluster, this feature causes Velero to back up _all_ Kubernetes API group versions that are supported on t ...
```
This introduction is too long and should be split into a separate concept module. The introduction needs to be a sentence or two that describes the action the user takes in the module. Please rework this module, after GA if necessary.

kalexand-rh commented on modules/oadp-using-enable-api-group-versions.adoc:
```diff
@@ -0,0 +1,49 @@
+
+[NOTE]
+====
+Enable API Group Versions is still in beta.
```
Is this supported?

RichardHoch replied:
Great question. @kaovilai ?

kaovilai replied:
From discussion it will be supported

kalexand-rh reviewed (commented):
Merge review for OADP GA. There are few things that need to be addressed before this can be merged.

RichardHoch commented on modules/oadp-using-enable-api-group-versions.adoc:
````diff
@@ -0,0 +1,49 @@
+* Configure the `EnableAPIGroupVersions` feature flag:
+
+[source,yaml]
+```
````
Done.

RichardHoch commented:
@kalexand-rh I made the changes you suggested and have asked @kaovilai whether this feature is supported.

mburke5678 commented:
/cherrypick enterprise-4.9

mburke5678 commented:
/cherrypick enterprise-4.10

mburke5678 commented:
/cherrypick enterprise-4.11

mburke5678 commented:
/cherrypick enterprise-4.12
